streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -1142,7 +1147,7 @@ void retryWithBackoff(final Map<TaskId, Set<TopicPartition>> tasksToBeCreated) {
                     }
                 }
 
-                if (tasksToBeCreated.isEmpty()) {
+                if (tasksToBeCreated.isEmpty() || backoffTimeMs > rebalanceTimeoutMs) {
```

Comment:
Sorry - i'm reading it wrong. I was failing to see that the backoff is done in the outer loop. Anyway, i still think the `backoffTimeMs` should be reset after each successful task creation. 

As for tracking the total rebalance time... we could just get the time at the start of `onPartitionsAssigned` and make sure that `MAX_POLL_INTERVAL_MS` hasn't passed since then.

Yes you probably could rely on the `CommitFailedException` to be thrown by poll. My concern is that we carry on for some reason and some tasks are never processed. In which case it would be better to find a way to bail early so they can be re-assigned.